### PR TITLE
[openocd] Work around problem with debug module by retrying failed commands

### PR DESCRIPTION
**Pull request by pamaury**
Workaround rv_dm bug in OpenOCD by retrying failed commands

See issue https://github.com/lowRISC/opentitan/issues/18087 for more details. In short, this commits works around the
problem by detecting when a DMI operations fails, wait for the DM/DMI to
get unstuck, reset the DMI and retry. It logs an ERROR so that problem does
not go unoticed.

The "DMI is stuck" condition is easily detectable because by requesting a
DMI reset, we get back the current value of DTMCS. This value is (per spec
and per implementation) always of the form 0x0000xyzt. When stuck, we will
get 0xffffffff which is clearly invalid. When that happens, we continue
issuing DMI resets until we see something different.

At the moment, and it is unclear why, we never get "partly invalid" values,
ie we either get a correct DTMCS value or 0xffffffff which is why the
detection above works. If it turns out that this assumption is wrong,
this patch might fail to detect the problem.

**Upstreaming??**: This fix might be upstreamed at least partially. It seems to me that OpenOCD does not correctly clears the error condition and should issue a DMI reset when the command fails (which it does not at the moment). This patch does more by retrying the command and this might not be accepted upstream.

### CODE REVIEW

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (andreaskurth) [inline third_party/openocd/reset_on_dmi_op_error.patch:16]: What does `ott` abbreviate in this context? I've seen it as abbreviation for opentitantool, but this has nothing to do with opentitantool, right?

(Also in the log messages below as `OTT`)

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (andreaskurth) [inline third_party/openocd/reset_on_dmi_op_error.patch:26]: Suggest to put on one line (`else break;`) or add parentheses

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (andreaskurth) [inline third_party/openocd/reset_on_dmi_op_error.patch:46]: Suggest to put on one line or add parentheses

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (andreaskurth) [inline third_party/openocd/reset_on_dmi_op_error.patch:57]: Suggest to put on one line or add parentheses

**Review by andreaskurth — COMMENTED** [review]
Thanks for implementing this workaround, @pamaury! I think this looks good overall, with some minor points

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (dmcardle) [inline third_party/openocd/reset_on_dmi_op_error.patch:23]: Should this sleep for some period and then `continue`? It looks like it might rapidly log this message 100 times.

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (pamaury) [reply third_party/openocd/reset_on_dmi_op_error.patch:16]: Yes I was thinking opentitantool because that's all the infrastructure we are using for testing but I can change the name if you want.

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (dmcardle) [inline third_party/openocd/reset_on_dmi_op_error.patch:27]: Optional nit: move this logic after the loop since it only is meant to run <= 1 times. I'd also make the loop `for (int try=0; try < kNumTries; try++)`.

**Review by dmcardle — APPROVED** [review]
LGTM! Thanks for your tenacious investigation, @pamaury!

In the interest of mitigating current flakiness, I support merging this now and upstreaming a patch to OpenOCD later. Due to the way we pull OpenOCD into Bazel, we'd need to wait for an official release/candidate anyway before we'd get to use the upstreamed patch in the OpenTitan repo. (We rely on the official SourceForge release archive to include OpenOCD's dependencies rather than fetching them ourselves.)

That being said, if hardware folks think this is fixable in OpenTitan or in rv_dm, that's probably the better solution.

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (pamaury) [reply third_party/openocd/reset_on_dmi_op_error.patch:26]: I am trying to follow the OpenOCD source code convention of that file (which is a bit flexible it seems). I am happy to change to OpenTitan's convention but I am not sure if this is the right approach.

Comment (pamaury) [reply third_party/openocd/reset_on_dmi_op_error.patch]: I have updated the patch.

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (pamaury) [reply third_party/openocd/reset_on_dmi_op_error.patch:23]: That's a good point. One tricky issue is that it's actually extremely rare that it stays stuck more than a couple of times. The maximum I have observed so far is 5 times, once (in 100s of runs). We do not understand the underlying issue real well and it could be that the jtag clocking done by doing a reset in a loop is important to unstuck the debug module. If that is the case then sleeping might actually not help.

Alternatively, I could keep a counter of failed attempts and print it once at the end of the loop?

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (pamaury) [reply third_party/openocd/reset_on_dmi_op_error.patch]: Ok will do.

File: third_party/openocd/reset_on_dmi_op_error.patch
Comment (andreaskurth) [reply third_party/openocd/reset_on_dmi_op_error.patch]: I see, thanks for changing it in the function name. I think it should also be changed in the log messages

**Review by timothytrippel — APPROVED** [review]
Thanks for the persistence in tracking this down and for the patch @pamaury !

I agree with @dmcardle , lets get this merged to unblock manufacturing testing and you can try to upstream the reset component on the side. I think that sounds reasonable from a "what to upstream" perspective.

**Comment by pamaury** [comment]
Here is an example [run of nightly](https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=113811&view=results) where the patch springs in action. See the artefact for the log, in particular `20230421.2/sw/device/silicon_creator/rom/e2e:rom_e2e_debug_test_otp_test_unlocked0.log`.

**Review by andreaskurth — APPROVED** [review]
Thanks for implementing and confirming this workaround, @pamaury!

**Comment by andreaskurth** [comment]
Now that we have identified and very likely fixed the root cause for the 'DMI is stuck' symptom, should we revert this PR and replace it with a solution where OpenOCD clears the error condition and resets the DMI when the command fails (ideally upstream first)? @pamaury  @timothytrippel

**Comment by timothytrippel** [comment]
That sounds good to me as well. Thanks everyone!

**Pull request merged**
